lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -718,3 +726,111 @@ instance
                 addresses (liftPaymentAddress @n @k) (externalPool s)
         in
             nonChangeAddresses <> changeAddresses
+
+--------------------------------------------------------------------------------
+--
+-- SeqAnyState
+--
+-- For benchmarking and testing arbitrary large sequential wallets.
+
+-- | An "unsound" alternative that can be used for benchmarking and stress
+-- testing. It re-uses the same underlying structure as the `SeqState` but
+-- it discover addresses based on an arbitrary ratio instead of respecting the
+-- BIP-44 discovery.
+--
+-- The proportion is stored as a type-level parameter so that we don't have to
+-- alter the database schema to store it. It simply exists and depends on the
+-- caller creating the wallet to define it.
+newtype SeqAnyState (network :: NetworkDiscriminant) key (p :: Nat) = SeqAnyState
+    { innerState :: SeqState network key
+    } deriving (Generic)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    , Show (k 'AddressK XPub)
+    , Show (KeyFingerprint "payment" k)
+    ) => Show (SeqAnyState n k p)
+
+instance
+    ( NFData (k 'AccountK XPub)
+    , NFData (k 'AddressK XPub)
+    , NFData (KeyFingerprint "payment" k)
+    )
+    => NFData (SeqAnyState n k p)
+
+-- | Initialize the HD random address discovery state from a root key and RNG
+-- seed.
+--
+-- The type parameter is expected to be a ratio of addresses we ought to simply
+-- recognize as ours. It is expressed in tenth of percents, so "1" means 0.1%,
+-- "10" means 1% and 1000 means 100%.
+mkSeqAnyState
+    :: forall (p :: Nat) n k.
+        ( SoftDerivation k
+        , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+        , MkKeyFingerprint k Address
+        , WalletKey k
+        , Bounded (Index (AddressIndexDerivationType k) 'AddressK)
+        )
+    => (k 'RootK XPrv, Passphrase "encryption")
+    -> AddressPoolGap
+    -> SeqAnyState n k p
+mkSeqAnyState credentials poolGap = SeqAnyState
+    { innerState = mkSeqStateFromRootXPrv credentials poolGap
+    }
+
+instance
+    ( SoftDerivation k
+    , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+    , KnownNat p
+    ) => IsOurs (SeqAnyState n k p) Address
+  where
+    isOurs (Address bytes) st@(SeqAnyState inner)
+        | crc32 bytes < p =
+            let
+                edge = Map.size (indexedKeys $ externalPool inner)
+                ix = toEnum (edge - fromEnum (gap $ externalPool inner))
+                pool' = extendAddressPool @n ix (externalPool inner)
+            in
+                (True, SeqAnyState (inner { externalPool = pool' }))
+        | otherwise =
+            (False, st)
+      where
+        p = floor (double (maxBound :: Word32) * double (natVal (Proxy @p)) / 1000)
```

Comment:
_Perhaps_ this is a bit easier to read:

```suggestion
        p = floor
            $ double (maxBound :: Word32)
            * double (natVal (Proxy @p))
            / 1000
```

Reply:
I disagree :), although I admit this could be more readable. What about:

```hs
        p = floor (double sup * double (natVal (Proxy @p)) / 1000)
          where
            sup = maxBound :: Word32 
```

Reply:
> What about: 
> ```haskell
>         p = floor (double sup * double (natVal (Proxy @p)) / 1000)
>           where
>             sup = maxBound :: Word32 
> ```
I think this is an improvement. I'm guessing `sup` is short for "superior" ("borne supérieure")?

Perhaps we could also give `natVal (Proxy @p)` a name, and then we'd have:
```hs
        p = floor ((double sup * double den) / 1000)
          where
            sup = maxBound :: Word32
            den = natVal (Proxy @p) 
```
Or:
```hs
        p = floor ((double n * double d) / 1000)
          where
            n = maxBound :: Word32
            d = natVal (Proxy @p) 
```
(where `n` and `d` are numerator and denominator, respectively).

Reply:
I agree, although I think the line is already very readable :)